packages/openapi-v3/src/decorators/parameter.decorator.ts

```diff
@@ -204,6 +204,30 @@ export namespace param {
      * @param name Parameter name.
      */
     password: createParamShortcut('query', builtinTypes.password),
+
+    /**
+     * Define a parameter accepting an object value encoded
+     * - as a JSON string, e.g. `filter={"where":{"id":1}}`); or
+     * - in multiple nested keys, e.g. `filter[where][id]=1`
+     *
+     * @param name Parameter name
+     * @param schema Optional OpenAPI Schema describing the object value.
+     */
+    object: function(
```

Comment:
I think we need to distinguish between two bits of information provided by `@param.query.deepObject`:
 - the value has type "object"
 - the value is encoded using `deepObject` style.

Object values provided in query can use the following encoding styles (see the section "[Style values](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.1.md#style-values)"):

 - deepObject (e.g. `?color[R]=100&color[G]=200&color[B]=150`)
 - form + `explode: true` (e.g. `?color=R,100,G,200,B,150`)
 - form + `explode: false` (e.g. `?R=100&G=200&B=150`)

As I see it, `@param.query.object` is saying that the value is of `object` type, similar how `@param.query.number` or `@param.query.boolean` works. To me, the fact that we allow `deepObject` style only, is a limitation we can lift later in the future. For example, we can allow callers to specify a different `style` and `explode` values by adding an optional third parameter:

```ts
@param.query.object('filter', FilterSchema, {style: 'deepObject'});
@param.query.object('filter', FilterSchema, {style: 'form', explode: false});
@param.query.object('filter', FilterSchema, {style: 'form', explode: true});
```

> While just in case people misuse it with parameters got from path(with style matrix or label), thought?

At the moment, `object` method is defined on `param.query` only. I agree we should eventually add `param.path.object` and `param.header.object` in the future, supporting encoding styles that are appropriate for those sources. To me, it still makes sense to use the same method name `object`, because the method name is describing the value type, not the encoding style.

Maybe I don't understand your concern, what kind of misuse do you have in mind?

Reply:
@bajtos 
> To me, the fact that we allow deepObject style only, is a limitation we can lift later in the future. 

ok I got it 👍 I think we are on the same page now, previously I thought you are creating a decorator particularly for `{style: 'deepObject'}`.

And my bad I mixed the usage of `param.query.object` and `param.object`.

Let's keep the name as `param.query.object`